### PR TITLE
[SPARKR-210] Some APIs in SQLContext

### PR DESCRIPTION
**Pull request by davies**
tables() and tableNames() is only available with latest spark-1.3

### CODE REVIEW

**Comment by shivaram** [comment]
Thanks @davies. This is pretty cool. Any reason you didn't add `registerDataFrameAsTable` ? That seems to be the only method missing from those listed in SPARKR-210. 

Also, a different question is after SPARKR-210 will we be missing any methods or is that list complete ?

**Comment by davies** [comment]
@shivaram Because we already have `registerTempTable(df, tableName)`, then `registerDataFrameAsTable(sqlCtx, df, tableName)` will not be useful, so I did not added it.

load() and loadExternalTable() will be added in another PR,  createDataFrame() will be in single PR, this is the difficult one.

That's all of them.

File: pkg/R/SQLContext.R [hunk 1]
```diff
@@ -92,6 +110,34 @@ table <- function(sqlCtx, tableName) {
   dataFrame(sdf) 
 }
 
+
+#' Tablesq
```
Comment (shivaram) [inline pkg/R/SQLContext.R]: Typo - extra `q` at the end

File: pkg/R/SQLContext.R [hunk 2]
```diff
@@ -102,11 +148,6 @@ table <- function(sqlCtx, tableName) {
 #' @export
 #' @examples
 #'\dontrun{
-#' sc <- sparkR.init()
```
Comment (shivaram) [inline pkg/R/SQLContext.R]: I don't think these examples need to be deleted. Its good style to have complete examples for each example that the user can run

File: pkg/NAMESPACE [hunk 3]
```diff
@@ -123,11 +123,15 @@ exportMethods("asc",
               "sumDistinct")
 
 export("jsonFile",
+       "jsonRDD",
```
Comment (shivaram) [inline pkg/NAMESPACE]: This is not related to this PR, but can we sort this alphabetically ?

File: pkg/inst/tests/test_sparkSQL.R [hunk 4]
```diff
@@ -20,6 +20,34 @@ test_that("jsonFile() on a local file returns a DataFrame", {
...
+test_that("test cache, uncache and clearCache", {
+  df <- jsonFile(sqlCtx, jsonPath)
+  registerTempTable(df, "temp")
+  cacheTable(sqlCtx, "temp")
+  uncacheTable(sqlCtx, "temp")
+  clearCache(sqlCtx)
+})
+
+test_that("test tableNames and tables", {
+  expect_true(length(tableNames(sqlCtx)) == 0)
```
Comment (shivaram) [inline pkg/inst/tests/test_sparkSQL.R]: Can we test this after registering a table ?

File: pkg/R/SQLContext.R (hunk 2 above)
Comment (davies) [inline pkg/R/SQLContext.R]: Make sense, will revert.

File: pkg/inst/tests/test_sparkSQL.R (hunk 4 above)
Comment (davies) [inline pkg/inst/tests/test_sparkSQL.R]: Good idea!

**Comment by shivaram** [comment]
Thanks for the clarification - I left some minor comments inline but otherwise LGTM.

For merging this we can either
(a) wait for the next RC
(b) change the build.sbt to 1.3.0-SNAPSHOT. I'll then I'll test locally and merge

**Comment by davies** [comment]
I tried to change it to 1.3.0-SNAPSHOT, but failed with:
```
[warn] 	Note: Unresolved dependencies path:
[warn] 		org.apache.spark:spark-core_2.10:1.3.0-SNAPSHOT (/Users/davies/work/SparkR-pkg/pkg/src/build.sbt#L23-54)
[warn] 		  +- edu.berkeley.cs.amplab:sparkr_2.10:0.1
[warn] 		org.apache.spark:spark-sql_2.10:1.3.0-SNAPSHOT (/Users/davies/work/SparkR-pkg/pkg/src/build.sbt#L23-54)
[warn] 		  +- edu.berkeley.cs.amplab:sparkr_2.10:0.1
sbt.ResolveException: unresolved dependency: org.apache.spark#spark-core_2.10;1.3.0-SNAPSHOT: not found
unresolved dependency: org.apache.spark#spark-sql_2.10;1.3.0-SNAPSHOT: not found
```

**Comment by shivaram** [comment]
Did you do `mvn install` or `sbt/sbt publish-local` from the Spark master branch on your laptop ? If it still doesn't work you can try to add the following to resolvers and that should do the trick 

```
"Local Maven Repository" at Path.userHome.asFile.toURI.toURL + ".m2/repository",
```

**Comment by cafreeman** [comment]
I don't think the library path normally looks like this, does it?

org.apache.spark#spark-core_2.10;1.3.0-SNAPSHOT

The '#' and ';' are throwing me off. Did you by any chance miskey something when you were updating build.sbt

**Comment by davies** [comment]
@shivaram I could build it locally, but how can we test it in travis after merging it?

**Comment by shivaram** [comment]
Ah I misunderstood -- We can't test it in travis with snapshot builds. I'll just test it locally on my machine and then merge.

**Comment by shivaram** [comment]
Also just to clarify since sparkr-sql is a development branch only developers working on this feature will be building it for now. Which is why I think its okay to depend on a snapshot build for a few days while we continue working on things.

**Comment by shivaram** [comment]
A slightly different solution in case we really want to make Travis happy. In sbt we can depend on a jar from any public URL. So I could for example upload a 1.3.0-SNAPSHOT jar to S3 and we can change out dependency to 

```
  "org.apache.spark" % "spark-core_2.10" % "1.3.0-SNAPSHOT" from "https://s3-us-west-1.amazonaws.com/sparkr-dep-1.3/2.10/spark-core_2.10-1.3.0-SNAPSHOT.jar"
```

Would anybody prefer that ?

**Comment by davies** [comment]
+1

**Comment by cafreeman** [comment]
I think I would. It's be nice to all be on the same page and also know that our test results will be consistent. Right now some of my unit tests fail on the snapshot because they're written for an "old" version of 1.3 and some of the data types have since changed.

**Comment by cafreeman** [comment]
@shivaram so is that Amazon link currently live? If so, is there also one for `spark-sql_2.10`?

**Comment by shivaram** [comment]
Not yet -- am looking at it now.

**Comment by shivaram** [comment]
Yeah so this is setup now. To use this you'll need to do the following -- In build.sbt change
```
-  val defaultSparkVersion = "1.3.0"
+  val defaultSparkVersion = "1.3.0-SNAPSHOT-sparkr"
```
and in resolvers add the S3 bucket 
```
 resolvers ++= Seq(
+  "SparkR 1.3 Snapshots" at "https://s3-us-west-2.amazonaws.com/sparkr-dep-1.3/release",
```

Can you test this and also send a PR with it if it works ? FWIW the build here is from Spark master branch at https://github.com/apache/spark/commit/2db6a853a53b4c25e35983bc489510abb8a73e1d

**Comment by cafreeman** [comment]
Building now. I'll let you now.

**Comment by cafreeman** [comment]
Actually, I'm working on a PR right now that already updates things to a new version (and fixes things like `Dsl` vs `functions`, etc). I'll submit that stuff together before submitting new methods.

**Comment by shivaram** [comment]
Thanks @davies for updating this. @cafreeman any more comments ?

File: pkg/R/SQLContext.R [hunk 5]
```diff
@@ -68,6 +99,7 @@ sql <- function(sqlCtx, sqlQuery) {
   dataFrame(sdf)
 }
 
+
```
Comment (cafreeman) [inline pkg/R/SQLContext.R]: nit: extra line

File: pkg/R/SQLContext.R [hunk 6]
```diff
@@ -92,6 +124,56 @@ table <- function(sqlCtx, tableName) {
   dataFrame(sdf) 
 }
 
+
```
Comment (cafreeman) [inline pkg/R/SQLContext.R]: nit: extra line

**Comment by cafreeman** [comment]
The code looks good, but @davies you may want to check for extra lines. I think you picked up a bunch of extras when you merged.

**Comment by davies** [comment]
@cafreeman These extra lines may be out of this PR, I think it's good to have two space line between each API (we use one between comment and API).

**Comment by shivaram** [comment]
Alright I'm merging this PR -- At some point we should revisit some of the style guidelines we are using in this project and (a) document it and (b) use an automatic style checker

I'll open a JIRA for that.

**Pull request merged**